### PR TITLE
Fixes unintended move of banner text, as reported in issue #8797.

### LINKED ISSUE #8797: C# namespace imports insertion point before fileheader
The current logic for importing new namespaces into C# code (I guess similar for VB but I didn't double check) from the class:

`roslyn/src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs
`
goes roughly like this:
1) Find existing imports and insert before
2) Or if there are none insert in the root

When using stylecop rules though the using statements are nested within the namespace, example:

```
<Current logic inserts here>
namespace MyProject
{
    <But should be inserting here?>
    using System;
}
```

But it also inserts new namespaces before file headers in this scenario since it doesn't find the other imports, example:

```
<Current logic inserts here>
// --------------------------------------------------------------------------------------------------------------------
// <copyright file="File.cs" company="MyOrgnaization">
// Copyright (C) MyOrgnaization 2016
// </copyright>
// --------------------------------------------------------------------------------------------------------------------
<If imports inside namespace are problematic, the at least skip the file header(s) and import here>
namespace MyProject
{
    <But should be ideally inserting here?>
    using System;
}
```

### PR DESCRIPTION
**Pull request by bkoelman**
My implementation was inspired by the logic in http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Organizing/Organizers/MemberDeclarationsOrganizer.cs.

Additionally I had to correct an existing broken test, which succeeded because some whitespace got stripped before.

Fixes #8797.

### CODE REVIEW

**Comment by bkoelman** [comment]
@Pilchie @CyrusNajmabadi @balajikris @brettfo @DustinCampbell could some of you please review?
/CC @markuskorbel

I would be great if this ships with Update2. Do I need to target another branch for that?

**Comment by davkean** [comment]
@dotnet-bot test eta please
@dotnet-bot test vsi please

**Comment by davkean** [comment]
tag @dotnet/roslyn-ide

**Comment by Pilchie** [comment]
@bkoelman In order to ship with VS Update 2, you would need to target `stabilization`.  However, in order to reduce the risk of introducing regressions, we're only approving fixes for issues that are significant enough into `stabilization` (mostly crashes, hangs, significant regressions, or totally broken features).  This issue does not meet that bar, so `master` is the right place for this fix. It will be in the next VS Update after Update 2.

**Comment by bkoelman** [comment]
@Pilchie That's understandable, thanks for the update. I'll see if I can somehow integrate the fix in my analyzer until the next VS Update ships.

**Comment by Pilchie** [comment]
@CyrusNajmabadi and/or @mattwar care to take a look here?

:+1: from me.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 1]
```diff
@@ -1217,6 +1231,23 @@ public override SyntaxNode InsertNamespaceImports(SyntaxNode declaration, int in
...
 
+        private static SyntaxList<UsingDirectiveSyntax> CopyBanner(CompilationUnitSyntax cu, SyntaxList<UsingDirectiveSyntax> usingsToInsert)
+        {
+            // First. Strip any pp directives or banners on the compilation unit.  They
+            // have to stay at the top of the list.
+            IEnumerable<SyntaxTrivia> banner = cu.GetLeadingBannerAndPreprocessorDirectives();
+
+            // Now, we want to remove any blank lines from the new first node and then reattach the banner. 
+            var finalFirstNode = usingsToInsert[0];
+            finalFirstNode = finalFirstNode.GetNodeWithoutLeadingBlankLines();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Why are you removing leading blank lines?

File: src/Workspaces/CSharpTest/CodeGeneration/AddImportsTests.cs [hunk 2]
```diff
@@ -354,14 +354,52 @@ class C
...
 
 class C 
 {
    public N.C F;
 }");
         }
+
+        [Fact]
+        [WorkItem(8797, "https://github.com/dotnet/roslyn/issues/8797")]
+        public async Task TestBannerTextRemainsAtTopOfDocument()
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharpTest/CodeGeneration/AddImportsTests.cs]: Do you have tests that show how leading whitespace lines are handled?

Comment (bkoelman) [inline src/Workspaces/CSharpTest/CodeGeneration/AddImportsTests.cs]: Added a test for that

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (bkoelman) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: I copied this from http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Organizing/Organizers/MemberDeclarationsOrganizer.cs,97.

Whitespace is removed from the old first node. Later, that leading whitespace is then added to the new first node (method InsertNamespaceImports at line 1221)

Comment (mattwar) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Why don't you use the PreservieTrivia pattern.  Maybe a PerserveBanner API with a callback with the CU with the banner text stripped, that gets reattached when the callback returns?

File: src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs [hunk 3]
```diff
@@ -70,12 +70,15 @@ protected override SyntaxNode InsertNamespaceImport(SyntaxNode root, SyntaxGener
                     : UsingsAndExternAliasesDirectiveComparer.NormalInstance;
 
             // find insertion point
+            int existingIndex = 0;
             foreach (var existingImport in gen.GetNamespaceImports(root))
             {
                 if (comparer.Compare(import, existingImport) < 0)
                 {
-                    return gen.InsertNodesBefore(root, existingImport, new[] { import });
```
Comment (mattwar) [inline src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs]: Whatever fix you do do make the banner trivia get preserved in the InsertNamespaceImports API should also be true when InsertNodesBefore is called, or ReplaceNode, etc.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (bkoelman) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: @mattwar Better?

File: src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs (hunk 3 above)
Comment (bkoelman) [inline src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs]: How to do that is not straightforward for me. Can you help out?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 4]
```diff
@@ -1202,21 +1202,31 @@ public override IReadOnlyList<SyntaxNode> GetNamespaceImports(SyntaxNode declara
 
         public override SyntaxNode InsertNamespaceImports(SyntaxNode declaration, int index, IEnumerable<SyntaxNode> imports)
```
Comment (mattwar) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: I meant making either making a function like PreserveBanner(SyntaxNode decl, Func<SyntaxNode, SyntaxNode> nodeChanger) or modify PreserveTrivia function to understand the banner..  If you modify PreserveTrivia then you get the behavior correctly everywhere.   Then you would change InsertNamespaceImports to call PreserveTrivia, which would then have an argument to another method InsertNamespaceImportsInternal, following the pattern of other Insert and With APIs.

Comment (bkoelman) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: @mattwar Thanks for explaining. I've added a commit. Is this what you'd like to see?

File: src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs (hunk 3 above)
Comment (bkoelman) [inline src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs]: It turns out that the existing downstream call to PreserveTrivia is sufficient for this to work. So I've reverted this change and added a test.

**Comment by mattwar** [comment]
Added issue #920 to enable VB fixer when SyntaxGenerator is fixed.

**Comment by Pilchie** [comment]
@mattwar - Can you take another look at this please?  I'd like to get it merged if possible.  It's getting pretty stale.

**Comment by mattwar** [comment]
:+1:

**Comment by Pilchie** [comment]
@brettfo Can you take a look at this vsi failure?

**Comment by brettfo** [comment]
VSI failure is a flaky test being tracked by #9671 and not related to this PR.

**Comment by Pilchie** [comment]
Thanks @brettfo, and thanks @bkoelman for the submission!

**Pull request merged**